lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -173,9 +205,10 @@ void ExecutionState::setResultCode(const ResultCode resultCode) {
 }
 
 ThreadPoolExecutor::~ThreadPoolExecutor() {
-  // Lock executionStatesMutex_ here to satisfy TSAN.
-  std::lock_guard<std::mutex> lock(executionStatesMutex_);
-  (void)lock;
+  // Wait for all inflight DeviceManager::runFunction() calls to return and be
+  // processed before starting to destroy state that is used in
+  // handleDeviceManagerResult().
+  inflightBarrier_.wait();
```

Comment:
You can avoid a race here by adding a `stopped` flag to the class which prevents further enqueues, this can return while there is an active call to run() which hasn't yet incremented the barrier. You should check that flag under the barrier's lock.

Reply:
The devil is in the details 😢

Reply:
Concurrency is hard ☹️